lib/src/authentication/token.dart

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import '../source/hosted.dart';
+
+/// Token is a structure for storing authentication credentials for third-party
+/// pub registries. A token holds registry [url], credential [kind] and [token]
+/// itself.
+///
+/// Token could be serialized into and from JSON format structured like
+/// this:
+///
+/// ```json
+/// {
+///   "url": "https://example.com/",
+///   "credential": {
+///     "kind": "Bearer",
+///     "token": "gjrjo7Tm2F0u64cTsECDq4jBNZYhco"
+///   }
+/// }
+/// ```
+class Token {
```

Comment:
@sigurdm suggests we call this a `Credential` (sorry, I probably suggest a name change to token).

But thinking about it, it makes sense because: it's a pair of _package repository_ and _secret token_.